moebot_bot/bot/commands/timer.go

```diff
@@ -23,115 +23,113 @@ type TimerCommand struct {
+					time:      time.Now(),
+					requestCh: nil,
 				}
-				chTimer.Unlock()
-			}
 
-			// Create a new timer
-			tc.chTimers.M[channelID] = &channelTimer{
-				time:      time.Now(),
-				writes:    0,
-				isWriting: false,
-				requestCh: make(chan string, 10),
+				pack.session.ChannelMessageSend(pack.message.ChannelID, "Timer started!")
+			} else if strings.EqualFold(pack.params[0], "stop") {
+				if chTimer, ok := tc.chTimers.M[channelID]; ok {
+					// Stop existing writer
+					if chTimer.requestCh != nil {
+						close(chTimer.requestCh)
+					}
+					delete(tc.chTimers.M, channelID)
+					pack.session.ChannelMessageSend(pack.message.ChannelID, "Timer stopped.")
+				} else {
+					pack.session.ChannelMessageSend(pack.message.ChannelID, "No timer running.")
+				}
 			}
 
 			tc.chTimers.Unlock()
-			pack.session.ChannelMessageSend(pack.message.ChannelID, "Timer started!")
 		} else {
 			pack.session.ChannelMessageSend(pack.message.ChannelID, pack.message.Author.Mention()+", you... you don't have permission to do that!")
 		}
 	} else {
-		tc.chTimers.RLock()
+		tc.chTimers.Lock()
 		if chTimer, ok := tc.chTimers.M[channelID]; ok {
-			chTimer.Lock()
-			// Reset the number of writes
-			chTimer.writes = 0
-
-			// If the time is not writing, start it
-			if !chTimer.isWriting {
-				go chTimer.writeTimes(pack)
-				chTimer.isWriting = true
+			// Close existing writer, then start a new one
+			if chTimer.requestCh != nil {
+				close(chTimer.requestCh)
 			}
-			chTimer.Unlock()
+			chTimer.requestCh = make(chan string, 10)
+			go writeTimes(pack, chTimer.time, chTimer.requestCh)
 		} else {
 			pack.session.ChannelMessageSend(pack.message.ChannelID, "No timer started for this channel...")
 		}
-		tc.chTimers.RUnlock()
+		tc.chTimers.Unlock()
 	}
 }
 
-func (ct *channelTimer) writeTimes(pack *CommPackage) {
-	duration := time.Since(ct.time)
+func writeTimes(pack *CommPackage, startTime time.Time, reqCh <-chan string) {
+	duration := time.Since(startTime)
+	writes := 0
+	currentWriteInterval := writeInterval
 
 	// Write the time once right away
 	go func() {
 		pack.session.ChannelMessageSend(pack.message.ChannelID, fmtDuration(duration))
-		ct.Lock()
-		ct.writes++
-		ct.Unlock()
 	}()
+	writes++
 
 	// Synchronize the writes to be divisible by the interval (works well when interval is 5 so we get writes at times like 0:30, 0:35, 0:40, etc.)
 	timeToSync := writeInterval - (duration % writeInterval)
 	time.Sleep(timeToSync)
 	duration += timeToSync
 
-	// Write again if we spent a sufficient time syncing, otherwise just wait until the next write interval
-	go func() {
-		if timeToSync > time.Second {
-			ct.Lock()
-			pack.session.ChannelMessageSend(pack.message.ChannelID, fmtDuration(duration))
-			ct.writes++
-			ct.Unlock()
-		}
-	}()
+	// Set the interval to a short duration if we are going to do an "after-sync" write so that messages sent to the receiver will be handled first, and then do the write shortly afterwards.
```

Comment:
Please limit lines to preferably <100 characters, but absolutely no more than 160